lib/motors/PID.cpp

```diff
@@ -0,0 +1,56 @@
+#include "PID.hpp"
+
+void PIDController::PIDController_Init()
+{
+  /* Clear controller variables */
+  integrator      = 0.0f;
+  prevError       = 0.0f;
+  differentiator  = 0.0f;
+  prevMeasurement = 0.0f;
+
+  out = 0.0f;
+};
+
+float PIDController::PIDController_Update(float setpoint, float measurement)
+{
+  // Error signal
+  float error = setpoint - measurement;
+
+  // Proportional term
+  float proportional = Kp * error;
+
+  // Integral term
+  integrator = integrator + 0.5f * Ki * T * (error + prevError);
+
+  // Integrator Anti wind-up (dynamic integrator clamping)
+  if (integrator > limMaxInt) {
+    integrator = limMaxInt;
+
+  } else if (integrator < limMinInt) {
+    integrator = limMinInt;
+  }
+
+  // Derivative Term (optional) with low pass filter
+  differentiator
+    = -(2.0f * Kd * (measurement - prevMeasurement) /* Note: derivative on measurement, therefore
+                                                       minus sign in front of equation! */
+        + (2.0f * tau - T) * differentiator)
+      / (2.0f * tau + T);
+
+  // Set controller output and limit outputs if needed
+  out = proportional + integrator + differentiator;
```

Comment:
to clarify what i said in the header, declare this here:
```c++
const core::Float output = ...
```

Reply:
This was also initialised to zero in what was `PIDController_Init` and is now the constructor `PidController`. Do I declare and initialise there? This kind of links to my previous question above.

Reply:
yeah, instead of making it a class data member you can just declare it on each iteration here, cos i don't believe you need the output value to persist between calls of `update(...)`